bindings/pybind11/idyntree_type_caster.h

```diff
@@ -0,0 +1,224 @@
+#ifndef IDYNTREE_PYBIND11_IDYNTREE_TYPE_CASTER_H
+#define IDYNTREE_PYBIND11_IDYNTREE_TYPE_CASTER_H
+
+#include <iostream>
+#include <string>
+#include <type_traits>
+
+#include <iDynTree/Core/MatrixDynSize.h>
+#include <iDynTree/Core/MatrixFixSize.h>
+#include <iDynTree/Core/Rotation.h>
+#include <iDynTree/Core/VectorDynSize.h>
+#include <iDynTree/Core/VectorFixSize.h>
+
+#include <pybind11/numpy.h>
+#include <pybind11/pybind11.h>
+
+namespace pybind11
+{
+namespace detail
+{
+template <template <unsigned int...> class base, typename derived>
+struct is_base_of_template_int_impl
+{
+    template <unsigned int... Ts> static constexpr std::true_type test(const base<Ts...>*);
+    static constexpr std::false_type test(...);
+    using type = decltype(test(std::declval<derived*>()));
+};
+
+template <template <unsigned int...> class base, typename derived>
+using is_base_of_template_int = typename is_base_of_template_int_impl<base, derived>::type;
+
+template <typename Type>
+using is_idyntree_vector_fix_size = is_base_of_template_int<iDynTree::VectorFixSize, Type>;
+template <typename Type>
+using is_idyntree_matrix_fix_size = is_base_of_template_int<iDynTree::MatrixFixSize, Type>;
+
+// Type caster for all the classes that inherits from VectorDynSize
```

Comment:
(And by consequence all the SFINAE-enabler logic above)

Why templating the caster? 
iDynTree vector is not a template class so I do not understand why you need all this logic.

Can you do one of:
- Add lots of comment on why the logic is here, or
- Remove the logic altogether. AFAIK there is no need for all of this, or
- Enable the SFINAE to match any of the 2 vector combination (i.e. Vector[Fix/Dyn]Size and Matrix[Fix|Dyn]Size) and have only 2 caster. You might also refactor some of the code to have more common code and avoid repetitions

Reply:
Before proceeding with the modification I would like to explain the rationale here. I implemented this to remove automatically convert all the objects that inherit from `Vector[Fix/Dyn]Size` and `Matrix[Fix|Dyn]Size`. 
Now I think this is not the right way to do it, right?

Reply:
Maybe the first question is: which objects currently inherit from the vector/matrix classes?

Second question. If you want to avoid those methods to be matched by derived classes, why not using directly `std::is_same`?

E.g.
```cpp
#include <iostream>
#include <type_traits>

class Base {
 public:
void bar() {
    std::cerr << "Base::bar" << std::endl;
}

};

class Derived: public Base {
 public:
    void bar() {
    std::cerr << "Derived::bar" << std::endl;
}
};

class Unrelated {
public:
    void bar() {
    std::cerr << "Unrelated::bar" << std::endl;
}
};


template <typename T>
void foo(T obj) {
    obj.bar();
}

template <typename T,
typename = std::enable_if_t<std::is_base_of_v<Base, T>>>
void fooOnBaseTree(T obj) {
    obj.bar();
}

template <typename T,
typename = std::enable_if_t<std::is_same_v<Base, T>>>
void fooOnBaseOnly(T obj) {
    obj.bar();
}

int main() {

    std::cerr << std::endl << "foo" << std::endl;
    foo(Base());
    foo(Derived());
    foo(Unrelated());

    std::cerr << std::endl << "fooOnBaseTree" << std::endl;
    fooOnBaseTree(Base());
    fooOnBaseTree(Derived());
    // No matching function for the next call.
    // fooOnBaseTree(Unrelated());

    std::cerr << std::endl << "fooOnBaseOnly" << std::endl;
    fooOnBaseOnly(Base());
    // No matching function for the next calls.
    // fooOnBaseOnly(Derived());
    // fooOnBaseOnly(Unrelated());

    return 0;
}
```

This will print
```

foo
Base::bar
Derived::bar
Unrelated::bar

fooOnBaseTree
Base::bar
Derived::bar

fooOnBaseOnly
Base::bar
```

Reply:
> Maybe the first question is: which objects currently inherit from the vector/matrix classes?


This is actually the problem. Indeed given the current implementation, several classes are automatically converted in NumPy vectors fox instance
`iDynTree::Direction` is bindings have been removed, The same is valid for `Axis`, `RotationalInertiaRaw`, `SpatialInertiaRaw` and `SpatialInertia`.

I think that this approach is wrong and only `iDynTree::Vectors` and `iDynTree::Matrices` should be automatically converted in NumPy objects. All the rest should remain as it is.

What do you think?

Reply:
Here https://github.com/robotology/idyntree/pull/931/commits/ecc64ad17c9f1018b8e9df5766673a3d66255c90 I implemented the third point sudgeted in https://github.com/robotology/idyntree/pull/931#discussion_r754973006